src/backend/distributed/utils/background_jobs.c

```diff
@@ -385,292 +823,150 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
-		 * update it to reflect the new state
-		 */
-		task->status = BACKGROUND_TASK_STATUS_RUNNING;
-		SET_NULLABLE_FIELD(task, pid, pid);
-
-		/* Update task status to indicate it is running */
-		UpdateBackgroundTask(task);
-		UpdateBackgroundJob(task->jobid);
-
+		AssignRunnableTasks(&queueMonitorExecutionContext);
 		PopActiveSnapshot();
 		CommitTransactionCommand();
 
-		MemoryContextSwitchTo(perTaskContext);
-
-		bool hadError = false;
-		StringInfoData message = { 0 };
-		initStringInfo(&message);
-
-		{
-			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
-										  dsm_segment_address(seg));
-			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
-			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
-
-			/*
-			 * We consume the complete shm_mq here as ConsumeTaskWorkerOutput loops till
-			 * it reaches a SHM_MQ_DETACHED response.
-			 */
-			ConsumeTaskWorkerOutput(responseq, &message, &hadError);
+		/* get running task entries from hash table */
+		List *runningTaskEntries = GetRunningTaskEntries(
+			queueMonitorExecutionContext.currentExecutors);
+		hasAnyTask = list_length(runningTaskEntries) > 0;
 
-			shm_mq_detach(responseq);
-		}
+		/* useful to sleep if all tasks ewouldblock on current iteration */
+		queueMonitorExecutionContext.allTasksWouldBlock = true;
 
+		/* monitor executors inside transaction */
 		StartTransactionCommand();
 		PushActiveSnapshot(GetTransactionSnapshot());
 
-		/*
-		 * Same as before, we need to lock pg_dist_background_task in a way where we can
-		 * check if there had been a concurrent cancel.
-		 */
-		LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
-
-		oldContext = MemoryContextSwitchTo(perTaskContext);
-		task = GetBackgroundTaskByTaskId(task->taskid);
-		MemoryContextSwitchTo(oldContext);
-
-		if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING ||
-			task->status == BACKGROUND_TASK_STATUS_CANCELLED)
+		/* iterate over all handle entries and monitor each task's output */
+		BackgroundExecutorHashEntry *handleEntry = NULL;
+		foreach_ptr(handleEntry, runningTaskEntries)
 		{
-			/*
-			 * A concurrent cancel has happened or the task has disappeared, we are not
-			 * retrying or changing state, we will only reflect the message onto the task
-			 * and for completeness we update the job aswell, this should be a no-op
-			 *
-			 * We still need to release the shared memory and xact before looping
-			 */
-
-			dsm_detach(seg);
-
-			task->status = BACKGROUND_TASK_STATUS_CANCELLED;
-			task->message = message.data;
-			UpdateBackgroundTask(task);
-			UpdateBackgroundJob(task->jobid);
+			/* create task execution context and assign it to queueMonitorExecutionContext */
+			TaskExecutionContext taskExecutionContext = {
+				.handleEntry = handleEntry,
+				.task = NULL,
+				.error = false
+			};
+			queueMonitorExecutionContext.taskExecutionContext = &taskExecutionContext;
```

Comment:
nit: I feel like conceptually this should be nested the other way around, i.e. queue monitor ctx should be inside a task execution context.

Reply:
sounds better.